File: src/main/java/rip/hippo/mosey/jar/impl/StandardJarLoader.java
/*
 * Mosey is a free and open source java bytecode obfuscator.
 *     Copyright (C) 2020  Hippo
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package rip.hippo.mosey.jar.impl;

import rip.hippo.mosey.asm.ClassHierarchy;
import rip.hippo.mosey.asm.wrapper.ClassWrapper;
import rip.hippo.mosey.jar.JarLoader;
import rip.hippo.mosey.jar.resource.ResourceManager;
import rip.hippo.mosey.jar.resource.impl.ClassResource;
import rip.hippo.mosey.jar.resource.impl.JarResource;
import rip.hippo.mosey.util.IOUtil;
import org.objectweb.asm.tree.ClassNode;
import org.tinylog.Logger;

import java.io.File;
import java.io.IOException;
import java.util.Enumeration;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

/**
 * @author Hippo
 * @version 1.0.0, 6/23/20
 * @since 1.0.0
 */
public final class StandardJarLoader implements JarLoader {

    @Override
    public void loadJar(File input, ResourceManager resourceManager, boolean library) {

        Logger.info(String.format("Loading %s " + (library ? "(library) " : ""), input.getAbsolutePath()));

        try (JarFile jarFile = new JarFile(input)) {
            Enumeration<JarEntry> entries = jarFile.entries();
            while (entries.hasMoreElements()) {
                JarEntry jarEntry = entries.nextElement();
                byte[] bytes = IOUtil.toByteArray(jarFile, jarEntry);
                Logger.info(String.format("Loading resource %s", jarEntry.getName()));

                if (library) {
                    if (jarEntry.getName().contains(".class")) {
                        registerHierarchy(new ClassResource(bytes, true));
                    }
                }else {
                    resourceManager.addResource(jarEntry.getName().contains(".class") ? registerHierarchy(new ClassResource(bytes, false)) : new JarResource(jarEntry.getName(), bytes));
                }
            }
        } catch (IOException e) {
            Logger.error(e, String.format("Failed to read jar file %s.", input.getAbsolutePath()));
        }

        Logger.info("Resources loaded.\n");

    }

    private ClassResource registerHierarchy(ClassResource classResource) {
        ClassWrapper classWrapper = classResource.getClassWrapper();
        ClassHierarchy.registerClass(classWrapper);
        return classResource;
    }
}

File: src/main/scala/rip/hippo/mosey/jar/impl/StandardJarLoader.scala
package rip.hippo.mosey.jar.impl

import java.io.{File, IOException}
import java.util.jar.JarFile

import rip.hippo.mosey.asm.ClassHierarchy
import rip.hippo.mosey.configuration.Configuration
import rip.hippo.mosey.jar.JarLoader
import rip.hippo.mosey.jar.resource.ResourceManager
import rip.hippo.mosey.jar.resource.impl.{ClassResource, JarResource}
import rip.hippo.mosey.logger.Logger
import rip.hippo.mosey.util.IOUtil


/**
 * @author Hippo
 * @version 1.1.0, 7/9/20
 * @since 1.0.0
 */
final class StandardJarLoader extends JarLoader {

  override def loadJar(input: File, resourceManager: ResourceManager, library: Boolean, configuration: Configuration): Unit = {
    val log = !library || (library && configuration.shouldLogLibraries)
    if (log) Logger.info(String.format("Loading %s " + (if (library) "(library) " else ""), input.getAbsolutePath))

    val exclude: String => Boolean = x => configuration.getExcluded.exists(exclude => x.startsWith(exclude))

    try {
      val jarFile = new JarFile(input)
      val entries = jarFile.entries()
      while (entries.hasMoreElements) {
        val jarEntry = entries.nextElement()
        val bytes = IOUtil.toByteArray(jarFile, jarEntry)
        val classFile = jarEntry.getName.contains(".class")
        if (log) Logger.info(String.format("Loading resource %s", jarEntry.getName))
        if (library) {
          if (classFile) {
            registerHierarchy(ClassResource(bytes, true, configuration.shouldInlineJSR))
          }
        } else {
          resourceManager.addResource(if (classFile && !exclude(jarEntry.getName)) registerHierarchy(ClassResource(bytes, false, configuration.shouldInlineJSR)) else JarResource(jarEntry.getName, bytes))
        }
      }
      jarFile.close()
    } catch {
      case e: IOException => Logger.error(e, String.format("Failed to read jar file %s", input.getAbsolutePath))
    }
    if (log) Logger.info("Resources loaded.\n")
  }

  def registerHierarchy(classResource: ClassResource): ClassResource = {
    val classWrapper = classResource.classWrapper
    ClassHierarchy.registerClass(classWrapper)
    classResource
  }
}
